app/gui/src/providers/backends.ts

```diff
@@ -0,0 +1,100 @@
+import { BackendType } from '#/services/Backend'
+import LocalBackend from '#/services/LocalBackend'
+import ProjectManager, {
+  ProjectManagerEvents,
+  Path as ProjectManagerPath,
+} from '#/services/ProjectManager'
+import RemoteBackend from '#/services/RemoteBackend'
+import HttpClient from '#/utilities/HttpClient'
+import { createContextStore } from '@/providers'
+import { GuiConfig } from '@/providers/guiConfig'
+import { ToValue } from '@/util/reactivity'
+import * as common from 'enso-common'
+import invariant from 'tiny-invariant'
+import { computed, markRaw, onScopeDispose, readonly, ref, toValue } from 'vue'
+import { GetText } from './text'
+
+function useBackends(
+  httpClient: HttpClient,
+  config: ToValue<GuiConfig>,
+  rootDirPath: ToValue<string | undefined>,
+  getText: GetText,
+) {
+  const projectManager = computed(() => {
+    const rootPath = toValue(rootDirPath)
+    if (rootPath == null) return
+    const cfg = toValue(config)
+    if (cfg.projectManagerUrl == null) return
+    const rootDirectory = ProjectManagerPath(rootPath)
+    if (rootDirPath && cfg.projectManagerUrl != null)
+      return new ProjectManager(cfg.projectManagerUrl, rootDirectory)
+    else return null
+  })
+  const localBackend = computed(() =>
+    projectManager.value ? markRaw(new LocalBackend(projectManager.value)) : null,
+  )
+  const remoteBackend = markRaw(new RemoteBackend(httpClient, console, getText))
+
+  const backendByCategory = (category: Category) =>
+    pickBackend(category, remoteBackend, localBackend.value)
+
+  const backendForProjectType = (projectType: BackendType) => {
+    switch (projectType) {
+      case BackendType.remote:
+        return remoteBackend
+      case BackendType.local: {
+        invariant(
+          localBackend.value,
+          'Attempted to get a local backend for local project, but no local backend was provided.',
+        )
+        return localBackend.value
+      }
+    }
+  }
+
+  const didLoadingProjectManagerFail = ref(false)
+  const onProjectManagerLoadingFailed = () => {
+    didLoadingProjectManagerFail.value = true
+  }
+  document.addEventListener(ProjectManagerEvents.loadingFailed, onProjectManagerLoadingFailed)
+  onScopeDispose(() =>
+    document.removeEventListener(ProjectManagerEvents.loadingFailed, onProjectManagerLoadingFailed),
+  )
+
+  const reconnectToProjectManager = () => {
+    didLoadingProjectManagerFail.value = false
+    localBackend.value?.reconnectProjectManager()
+  }
+  return {
+    localBackend,
+    remoteBackend,
+    backendByCategory,
+    backendForProjectType,
+    didLoadingProjectManagerFail: readonly(didLoadingProjectManagerFail),
+    reconnectToProjectManager,
+  }
+}
+
+export const [provideBackends, injectBackends] = createContextStore('backends', useBackends)
+
+/**
+ * Pick the backend for the given category.
+ * @throws {Error} when a Local Backend is requested for a non-local project.
+ */
+export function pickBackend(
```

Comment:
Once previous comments are applied, this function should be unused and removed (probably not marked as such because of `export`).